casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -95,44 +96,33 @@ sealed abstract class MultiParentCasperInstances {
               else Capture[F].capture { blockBuffer += b }
         } yield ()
 
-      def contains(b: BlockMessage): F[Boolean] = Capture[F].capture {
-        blockLookup.contains(b.blockHash)
-      }
+      def contains(b: BlockMessage): F[Boolean] =
+        Capture[F].capture {
+          blockLookup.contains(b.blockHash)
+        }
 
       def deploy(d: Deploy): F[Unit] =
-        Capture[F].capture {
-          deployBuff += d
-          deployHist += d
-        } *> Log[F].info(s"CASPER: Received deploy $d")
+        for {
+          _ <- Capture[F].capture {
+                deployBuff += d
+                deployHist += d
+              }
+          _ <- Log[F].info(s"CASPER: Received deploy $d")
+        } yield ()
 
       def estimator: F[IndexedSeq[BlockMessage]] =
-        scoresMap.map(scores => {
-
-          @tailrec
-          def sortChildren[A <: (BlockHash) => Int](blks: IndexedSeq[BlockHash],
-                                                    scores: A): IndexedSeq[BlockHash] = {
-            val newBlks = blks
-              .flatMap(b => {
-                val empty                         = new mutable.HashSet[BlockHash]()
-                val c: mutable.HashSet[BlockHash] = _childMap.getOrElse(b, empty)
-                if (c.nonEmpty) {
-                  c.toIndexedSeq
-                } else {
-                  IndexedSeq(b)
-                }
-              })
-              .distinct
-              .sortBy(scores)(decreasingOrder)
-
-            if (newBlks == blks) {
-              blks
-            } else {
-              sortChildren(newBlks, scores)
-            }
-          }
-
-          sortChildren(IndexedSeq(genesis.blockHash), scores).map(blockLookup)
-        })
+        Capture[F].capture {
+          // TODO: Push up immutable boundary
```

Comment:
I am concerned about switching entirely to using immutable HashSets and HashMaps because I think the Scala implementation is not very efficient when it comes to updating. By which I mean that I don't think much of the structure is reused, which leads to a lot of copying and a lot of garbage to be collected by the JVM. I think that as we switch to an entirely functional implementation we might want to consider using other data structures instead.

Reply:
Sounds good. What data structure do you recommend? Hopefully we don't have to roll our own.

@henrytill

Reply:
Resolved through chat